parquet-variant/src/variant.rs

```diff
@@ -0,0 +1,719 @@
+    ) -> Result<impl Iterator<Item = Result<&'m str, ArrowError>>, ArrowError> {
+        let iterator = self
+            .offsets()
+            .map(move |offset_range| self.get_field_by_offset(offset_range?));
+        Ok(iterator)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct VariantObject<'m, 'v> {
+    pub metadata: &'m VariantMetadata<'m>,
+    pub value: &'v [u8],
+}
+impl<'m, 'v> VariantObject<'m, 'v> {
+    pub fn fields(&self) -> Result<impl Iterator<Item = (&'m str, Variant<'m, 'v>)>, ArrowError> {
+        todo!();
+        #[allow(unreachable_code)] // Just to infer the return type
+        Ok(vec![].into_iter())
+    }
+    pub fn field(&self, _name: &'m str) -> Result<Variant<'m, 'v>, ArrowError> {
+        todo!()
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct VariantArray<'m, 'v> {
+    pub metadata: &'m VariantMetadata<'m>,
+    pub value: &'v [u8],
+}
+
+impl<'m, 'v> VariantArray<'m, 'v> {
+    pub fn len(&self) -> usize {
+        todo!()
+    }
+
+    pub fn values(&self) -> Result<impl Iterator<Item = Variant<'m, 'v>>, ArrowError> {
+        todo!();
+        #[allow(unreachable_code)] // Just to infer the return type
+        Ok(vec![].into_iter())
+    }
+
+    pub fn get(&self, index: usize) -> Result<Variant<'m, 'v>, ArrowError> {
+        // The 6 first bits to the left are the value_header and the 2 bits
+        // to the right are the basic type, so we shift to get only the value_header
+        let value_header = first_byte_from_slice(self.value)? >> 2;
+        let is_large = (value_header & 0x04) != 0; // 3rd bit from the right
+        let field_offset_size_minus_one = value_header & 0x03; // Last two bits
+        let offset_size = OffsetSizeBytes::try_new(field_offset_size_minus_one)?;
+        // The size of the num_elements entry in the array value_data is 4 bytes if
+        // is_large is true, otherwise 1 byte.
+        let num_elements_size = match is_large {
+            true => OffsetSizeBytes::Four,
+            false => OffsetSizeBytes::One,
+        };
+        // Skip the header byte to read the num_elements
+        // The size of the num_elements entry in the array value_data is 4 bytes if
+        // is_large is true, otherwise 1 byte.
+        let num_elements = num_elements_size.unpack_usize(self.value, 1, 0)?;
+        let first_offset_byte = 1 + num_elements_size as usize;
+
+        let overflow =
+            || ArrowError::InvalidArgumentError("Variant value_byte_length overflow".into());
+
+        // 1.  num_elements + 1
+        let n_offsets = num_elements.checked_add(1).ok_or_else(overflow)?;
+
+        // 2.  (num_elements + 1) * offset_size
+        let value_bytes = n_offsets
+            .checked_mul(offset_size as usize)
+            .ok_or_else(overflow)?;
+
+        // 3.  first_offset_byte + ...
+        let first_value_byte = first_offset_byte
+            .checked_add(value_bytes)
+            .ok_or_else(overflow)?;
```

Comment:
nit: Any particular reason this is spread out, instead of one chain like `Metadata::try_new` did?
```suggestion
        // (num_elements + 1) * offset_size + first_offset_byte
        let n_offsets = num_elements
            .checked_add(1)
            .and_then(|n| n.checked_mul(offset_size as usize))
            .and_then(|n| n.checked_add(value_bytes))
            .ok_or_else(overflow)?;
```